### DISCUSSION TITLE
调用QCefDownloadItem::start(path, true)未弹出文件保存对话框

### DISCUSSION THREAD

**Discussion by lxylbds** [CefView/QCefView#290] [Q&A]
cef使用当前支持的113，代码用的master分支，保存文件对话框未弹出，path尝试用:"","C:\test\dir1\file.log","C:\\\\test\\\\dir1\\\\file.log"都不灵，但是上传文件对话框可以弹出，请问是什么原因？
qt版本试了5.15.2及6.5.0，都无法弹出。用QCefViewTest同样无法弹出。
断点调试，问题出在download_item_callback_cpptoc.cc文件中，但是原码里木有这个文件，所以看不到详情

**Comment by tishion** [comment]
我用了最新版，暂时无法复现你说的问题，

另外需要注意的
```cpp
  /// <summary>
  /// Starts to download the item.
  /// </summary>
  /// <param name="path">The full path name (must include file name) to save the downloaded item</param>
  /// <param name="useDefaultDialog">Whether to use the default 'Save As...' dialog or not</param>
  /// <remarks>
  /// The 'path' parameter only works when 'useDefaultDialog' is set to false.
  /// If you set 'useDefaultDialog' to true then you cannot control the initial
  /// locatio nof the opened 'Save As...' dialog, it is determined by CEF internal implementation.
  /// </remarks>
  void start(const QString& path, bool useDefaultDialog = true) const;
```

1. path只有在useDefaultDialog=false的情况下才生效，直接默认下载到指定路径
2. 如果useDefaultDialog=true，打开的对话库的初始目录是CEF内部控制

**Reply by lxylbds** [reply]
useDefaultDialog=false也试了，传入了path，文件并未下载

**Comment by L-Super** [comment]
相关的源代码可否贴出来？

**Comment by man345** [comment]
我也碰到了这个问题，楼主解决了吗

**Comment by L-Super** [comment]
> 我也碰到了这个问题，楼主解决了吗
> 

目前已经实现了文件对话框。请检查代码是否为最新

**Reply by man345** [reply]
使用的是最新的，从master上下载的。
使用方式如下：
构建了一个MyCefView类，继承QCefView，在MyCefView中重写了onNewDownloadItem函数，在该函数中调用item->start()，没有对话框弹出。
![image](https://github.com/CefView/QCefView/assets/81149223/07c87d86-8691-4837-990a-acfc55124cac)

**Reply by man345** [reply]
是使用的方式不对吗

**Reply by man345** [reply]
使用库版本来源
![image](https://github.com/CefView/QCefView/assets/81149223/56094a66-307c-461b-8f74-f8ed2e46a8b0)

**Reply by L-Super** [reply]
尝试运行QCefViewTest demo下载文件测试，如果能正常弹出，则是你的实现有问题

**Reply by man345** [reply]
原因有应该是：
onNewDownloadItem(const QSharedPointer< QCefDownloadItem >&item, const QString & suggestedName)中形参item是一个共享指针的引用，此时传入的计数引用计数可能就是1，如果之间在函数onNewDownloadItem中直接使用item->start()，在函数作用域结束后，会导致item被释放，所以无法弹出对话框，所以需要新增一个共享指针去指向item，保证下载动作执行完成后，再释放item指向的内存空间。

**Reply by L-Super** [reply]
你的用法或许是错误的，请参考：https://github.com/CefView/QCefView/blob/c05b1234cd7199dd4e56cf79af10fb3f51e50755/example/QCefViewTest/DownloadManager.cpp#L12C1-L26C2

**Reply by qidongdong** [reply]
> 原因有应该是： onNewDownloadItem(const QSharedPointer< QCefDownloadItem >&item, const QString & suggestedName)中形参item是一个共享指针的引用，此时传入的计数引用计数可能就是1，如果之间在函数onNewDownloadItem中直接使用item->start()，在函数作用域结束后，会导致item被释放，所以无法弹出对话框，所以需要新增一个共享指针去指向item，保证下载动作执行完成后，再释放item指向的内存空间。

我也遇到了该情况 item生命周期不是根本问题
1CefLifeSpanHandler  onBeforePopup 进入 onBeforeNewBrowserCreate 此时是新的QCefView实例 并没有重载onNewDownloadItem 和
onUpdateDownloadItem 里面是空的 并没有调用start函数 触发 d_ptr->beforeDownloadCallback->Continue(path.toStdString(), useDefaultDialog);
我的解决方案是 比较粗糙 只是为了快速解决问题https://github.com/CefView/QCefView/blob/c05b1234cd7199dd4e56cf79af10fb3f51e50755/example/QCefViewTest/DownloadManager.cpp#L12C1-L26C2 实现放到QCefView里面认为这是默认行为。
